### PR TITLE
UriValidation implementation

### PR DESCRIPTION
**Pull request by jvnaiduwin**
Urivalidator Implementation.

This will validate Uri on basis on the Uri passed, whether its valid or not and  return boolean or UStatus.

Note : 8 test case differ in result when comparing to Java.

### CODE REVIEW

File: include/uprotocol-cpp/uri/validator/UriValidator.h [hunk 1]
```diff
@@ -19,21 +19,92 @@
...
+/* UriValidator class to validate uri*/
+    class UriValidator {
+        public:
 
-bool valid_uri(const std::string& uri) {
-    auto uri_view = LongUriSerializer::deserialize(uri);
-    return !isEmpty(uri_view);
-}
+         /**
+          * isValidUri a {@link uri} to ensure that it has at least a name for the
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/validator/UriValidator.h:41]: Use @brief on all of these.
Indicate [in], [out], or [in/out] for each parameter for all of these.

File: include/uprotocol-cpp/uri/validator/UriValidator.h [hunk 2]
```diff
@@ -19,21 +19,92 @@
  * under the License.
  * 
  * SPDX-FileType: SOURCE
- * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-FileCopyrightText: 2024 General Motors GTO LLC
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/validator/UriValidator.h:22]: Copyright doesn't lose original start date: Change to: 2023-2024

File: include/uprotocol-cpp/cloudevent/factory/cloud_event_factory.h [hunk 3]
```diff
@@ -469,7 +469,7 @@ struct factory {
       (*(ce).mutable_attributes())[Serializer::REQ_ID_KEY] = *attr;
     }
 
-    if (unlikely(rpcUri.empty() || !valid_uri(rpcUri))) {
+    if (unlikely(rpcUri.empty() || !UriValidator::isValidUri(rpcUri))) {
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/cloudevent/factory/cloud_event_factory.h:472]: We should not be using any C++20 constructs.  Remove all "likely" and "unlikely"'s.

File: include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h [hunk 4]
```diff
@@ -139,7 +139,7 @@ class MicroUriSerializer {
      * @param resource 
      * @return 
      */
-    [[nodiscard]] [[maybe_unused]] auto isMicroForm(const uprotocol::v1::UResource &resource) -> bool {
+    [[nodiscard]] [[maybe_unused]] static auto isMicroForm(const uprotocol::v1::UResource &resource) -> bool {
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h:142]: Remove all [[nodiscard]] and [[maybe_unusued]].

File: include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h [hunk 5]
```diff
@@ -148,7 +148,7 @@ class MicroUriSerializer {
       * @param resource 
       * @return 
       */
-    [[nodiscard]] [[maybe_unused]] auto isMicroForm(const uprotocol::v1::UEntity &entity) -> bool {
+    [[nodiscard]] [[maybe_unused]] static auto isMicroForm(const uprotocol::v1::UEntity &entity) -> bool {
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h:151]: Do not use "auto" except for obtaining a function return.  Please use the known type of bool.  This happens many times in this file.  Remove "-> bool".  No need for lambda here.

File: include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h [hunk 6]
```diff
@@ -148,7 +148,7 @@ class MicroUriSerializer {
...
         return entity.has_id() && entity.id() > 0;
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h:152]: Please enclose logical operation in separate parenthesis.  Many other instances of this in this file also.

File: include/uprotocol-cpp/uri/validator/UriValidator.h [hunk 7]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023 General Motors GTO LLC
+ * Copyright (c) 2024 General Motors GTO LLC
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/validator/UriValidator.h:2]: Return start year 2023-2024.

File: include/uprotocol-cpp/uri/validator/UriValidator.h [hunk 8]
```diff
@@ -19,21 +19,92 @@
...
 
 #include <uprotocol-cpp/uri/serializer/LongUriSerializer.h>
 #include <uprotocol-cpp/uri/tools/Utils.h>
+#include "uprotocol/uri.pb.h"
+#include "uprotocol/ustatus.pb.h"
+#include <string.h>
 
+using namespace uprotocol::v1;
 namespace uprotocol::uri {
+/* UriValidator class to validate uri*/
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/validator/UriValidator.h:36]: Go ahead and make this an @brief Doxygen comment.

File: include/uprotocol-cpp/uri/validator/UriValidator.h [hunk 9]
```diff
@@ -19,21 +19,92 @@
...
-bool valid_uri(const std::string& uri) {
-    auto uri_view = LongUriSerializer::deserialize(uri);
-    return !isEmpty(uri_view);
-}
+         /**
+          * isValidUri a {@link uri} to ensure that it has at least a name for the
+          * uEntity.
+          *
+          * @param uri {@link uri} to validate.
+          * @return Returns bool containing a true or a false
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/validator/UriValidator.h:45]: Indicate which is true (I would assume that valid is true)

File: include/uprotocol-cpp/uri/validator/UriValidator.h [hunk 10]
```diff
@@ -19,21 +19,92 @@
...
+          */
+
+          static bool isValidUri(const std::string& uri);
+
+         /**
+          * Validate a {@link UUri} to ensure that it has at least a name for the
+          * uEntity.
+          *
+          * @param uri {@link UUri} to validate.
+          * @return Returns Status containing a success or a failure with the error
```
Comment (MarkStokes71) [inline include/uprotocol-cpp/uri/validator/UriValidator.h:55]: Go ahead and put the end of this sentence on the same line.  Lines should be less than 120 characters.  Several other places in this file.

File: src/uri/LongUriSerializer.cpp [hunk 11]
```diff
@@ -65,7 +65,7 @@ auto uprotocol::uri::LongUriSerializer::serialize(const v1::UUri& uri) -> std::s
  * @return Returns an UUri data object.
  */
 auto uprotocol::uri::LongUriSerializer::deserialize(std::string const& protocol_uri) -> v1::UUri {
-    if (protocol_uri.empty()) {
+    if (nullptr == &protocol_uri ||  protocol_uri.empty()) {
```
Comment (MarkStokes71) [inline src/uri/LongUriSerializer.cpp:68]: Surround logical expressions in parentesis.
For example:  ((nullptr == &protocol_uri) ||  protocol_uri.empty())

File: src/uri/UriValidator.cpp [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+#include "uprotocol/ustatus.pb.h"
+#include <uprotocol-cpp/uri/serializer/LongUriSerializer.h>
+#include <uprotocol-cpp/uri/serializer/MicroUriSerializer.h>
+#include <string.h>
+
+using namespace uprotocol::uri;
+using namespace uprotocol::v1;
+
+/**
+ * isValidUri a {@link uri} to ensure that it has at least a name for the
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:37]: Add @brief for all brief method descriptions.

File: src/uri/UriValidator.cpp [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+ * uEntity.
+ *
+ * @param uri {@link uri} to validate.
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:40]: Add in/out as described above to all parameters.

File: src/uri/UriValidator.cpp [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+/**
+ * Validate a {@link UUri} to ensure that it has at least a name for the uEntity.
+ *
+ * @param uri {@link UUri} to validate.
+ * 
+ * @return Returns Status containing a success or a failure with the error message.
+ */
+UStatus UriValidator::validate(const UUri& uri) {
+  UStatus status;
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:56]: Indents shall be 4 spaces each.

File: src/uri/UriValidator.cpp [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+  if ((nullptr == &uri) || isEmpty(uri)) {
+  status.set_code(UCode::INVALID_ARGUMENT);
+    status.set_message("Uri is empty.");
+    return status;
+  }
+  if (uri.has_authority() && !isRemote(uri.authority())) {
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:62]: change these to "} else if {" and remove the return from each case.

File: src/uri/UriValidator.cpp [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+    status.set_code(UCode::INVALID_ARGUMENT);
+    status.set_message("Uri is remote missing uAuthority.");
+    return status;
+  }
+  if (!isEmpty(uri.entity())  && isBlank( uri.entity().name())) {
+    status.set_code(UCode::INVALID_ARGUMENT);
+    status.set_message("Uri is missing uSoftware Entity name.");
+    return status;
+  }
+  status.set_code(UCode::OK);
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:72]: This then would be the else case.  This allows the function to have only one return which is favorable for readability.

File: src/uri/UriValidator.cpp [hunk 17]
```diff
@@ -0,0 +1,159 @@
...
+ */
+UStatus UriValidator::validateRpcMethod(const UUri& uri) {
+    UStatus status = UriValidator::validate(uri);
+    if (status.code() != UCode::OK) {
+        return status;
+    }
+    if (!UriValidator::isRpcMethod(uri)) {
+      status.set_code(UCode::INVALID_ARGUMENT);
+      status.set_message("Invalid RPC method uri. Uri should be the method to be called, or method from response.");
+      return status;
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:91]: Remove this redundant return.

File: src/uri/UriValidator.cpp [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+ */
+UStatus UriValidator::validateRpcResponse(const UUri& uri) {
+    UStatus status = UriValidator::validate(uri);
+    if (status.code() != UCode::OK) {
+        return status;
+    }
+    if (!UriValidator::isRpcResponse(uri)) {
+        status.set_code(UCode::INVALID_ARGUMENT);
+        status.set_message("Invalid RPC response type.");
+        return status;
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:110]: Remove this redundant return.

File: src/uri/UriValidator.cpp [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+
+/**
+ * Returns true if URI is of type RPC.
+ *
+ * @param uri {@link UUri} to check if it is of type RPC method
+ * @return Returns true if URI is of type RPC.
+ */
+
+bool UriValidator::isRpcMethod(const UUri& uri) {
+    return (nullptr != &uri &&
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:124]: Enclose logical items with their own set of parenthesis. Align with logical operators on the left:
return ((nullptr != &uri)
       && (!isEmpty(uri.resource())
       && (uri.resource().has_instance())
and so on.

File: src/uri/UriValidator.cpp [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+ *
+ * @param uri {@link UUri} to check response
+ * @return Returns true if URI is of type RPC response.
+ */
+bool UriValidator::isRpcResponse(const UUri& uri) {
+  if (nullptr == &uri)
+  {
+    return false;
+  }
+  return (uri.resource().name().find("rpc") &&
```
Comment (MarkStokes71) [inline src/uri/UriValidator.cpp:156]: Same comment on return in isRpcMethod() above.

File: test/CMakeLists.txt [hunk 21]
```diff
@@ -108,6 +108,12 @@ target_compile_options(MicroUriSerializerTest PRIVATE -g -O0)
 target_link_libraries(MicroUriSerializerTest uprotocol-cpp gtest gtest_main)
 add_test("t-14-MicroUriSerializerTest"  MicroUriSerializerTest)
 
+add_executable(UriValidatorTest
```
Comment (MarkStokes71) [inline test/CMakeLists.txt:111]: Make this one line.

**Review by MarkStokes71 — COMMENTED** [review]
Good work.

File: include/uprotocol-cpp/uri/validator/UriValidator.h (hunk 1 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/validator/UriValidator.h:41]: Done

File: include/uprotocol-cpp/uri/validator/UriValidator.h (hunk 2 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/validator/UriValidator.h:22]: Done

File: include/uprotocol-cpp/cloudevent/factory/cloud_event_factory.h (hunk 3 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/cloudevent/factory/cloud_event_factory.h:472]: Done

File: include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h (hunk 4 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h:142]: remove  [[nodiscard]] and [[maybe_unusued]] from isMicroForm

File: include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h (hunk 5 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h:151]: Sure @MarkStokes71, i will convey the info to @MelamudMichael.

File: include/uprotocol-cpp/uri/validator/UriValidator.h (hunk 7 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/validator/UriValidator.h:2]: Done

File: include/uprotocol-cpp/uri/validator/UriValidator.h (hunk 9 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/validator/UriValidator.h:45]: Done

File: include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h (hunk 6 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/serializer/MicroUriSerializer.h:152]: Done

File: include/uprotocol-cpp/uri/validator/UriValidator.h (hunk 8 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/validator/UriValidator.h:36]: Done

File: include/uprotocol-cpp/uri/validator/UriValidator.h (hunk 10 above)
Comment (jvnaiduwin) [reply include/uprotocol-cpp/uri/validator/UriValidator.h:55]: Done

File: src/uri/LongUriSerializer.cpp (hunk 11 above)
Comment (jvnaiduwin) [reply src/uri/LongUriSerializer.cpp:68]: Done

File: src/uri/UriValidator.cpp (hunk 13 above)
Comment (jvnaiduwin) [reply src/uri/UriValidator.cpp:40]: Done

File: src/uri/UriValidator.cpp (hunk 12 above)
Comment (jvnaiduwin) [reply src/uri/UriValidator.cpp:37]: Done

File: src/uri/UriValidator.cpp (hunk 15 above)
Comment (jvnaiduwin) [reply src/uri/UriValidator.cpp:62]: Done

File: src/uri/UriValidator.cpp (hunk 17 above)
Comment (jvnaiduwin) [reply src/uri/UriValidator.cpp:91]: Done

File: src/uri/UriValidator.cpp (hunk 16 above)
Comment (jvnaiduwin) [reply src/uri/UriValidator.cpp:72]: Done

File: src/uri/UriValidator.cpp (hunk 19 above)
Comment (jvnaiduwin) [reply src/uri/UriValidator.cpp:124]: Done

File: src/uri/UriValidator.cpp (hunk 20 above)
Comment (jvnaiduwin) [reply src/uri/UriValidator.cpp:156]: Done

File: test/CMakeLists.txt (hunk 21 above)
Comment (jvnaiduwin) [reply test/CMakeLists.txt:111]: Done

**Review by jvnaiduwin — COMMENTED** [review]
Please review

**Pull request closed without merging**